sdk/compatability/src/env.rs

```diff
@@ -0,0 +1,453 @@
+        Self {
+            json_rpc_url,
+            coffer,
+            chain_id,
+        }
+    }
+
+    pub fn json_rpc_url(&self) -> &str {
+        &self.json_rpc_url
+    }
+
+    pub fn client(&self) -> BlockingClient {
+        BlockingClient::new(self.json_rpc_url())
+    }
+
+    pub fn coffer(&self) -> &Coffer {
+        &self.coffer
+    }
+
+    pub fn chain_id(&self) -> ChainId {
+        self.chain_id
+    }
+
+    pub fn transaction_factory(&self) -> TransactionFactory {
+        TransactionFactory::new(self.chain_id())
+    }
+}
+
+pub enum Coffer {
+    Faucet(FaucetClient),
+    MintKey(Ed25519PrivateKey),
+}
+
+impl Coffer {
+    pub fn fund(&self, currency: Currency, auth_key: AuthenticationKey, amount: u64) -> Result<()> {
+        match self {
+            Coffer::Faucet(faucet) => faucet.fund(currency, auth_key, amount),
+            Coffer::MintKey(_) => todo!(),
+        }
+    }
+}
+
+pub struct FaucetClient {
+    url: String,
+    json_rpc_client: BlockingClient,
+}
+
+impl FaucetClient {
+    pub fn new(url: String, json_rpc_url: String) -> Self {
+        Self {
+            url,
+            json_rpc_client: BlockingClient::new(json_rpc_url),
+        }
+    }
+
+    pub fn fund(&self, currency: Currency, auth_key: AuthenticationKey, amount: u64) -> Result<()> {
+        let client = reqwest::blocking::Client::new();
+
+        let url = format!(
+            "{}?amount={}&auth_key={}&currency_code={}&return_txns=true",
+            self.url,
+            amount,
+            auth_key,
+            currency.as_str(),
+        );
+
+        //TODO Fix faucet
+        // The current implementation of Faucet is racy, in that 2 independent requests can come in
+        // and result in faucet sending 2 txns to the network with the same sequence number and
+        // they race to complete. One request will complete while the other will get a HashMismatch
+        // Error.
+        'faucet: loop {
+            let response = client.post(&url).send()?;
+            let status_code = response.status();
+            let body = response.text()?;
+            if !status_code.is_success() {
+                return Err(format_err!(
+                    "Failed to query remote faucet server[status={}]: {:?}",
+                    status_code.as_str(),
+                    body,
+                ));
+            }
+            let bytes = hex::decode(body)?;
+            let txns: Vec<SignedTransaction> = bcs::from_bytes(&bytes).unwrap();
+
+            for txn in txns {
+                let txn_view = match self
+                    .json_rpc_client
+                    .wait_for_signed_transaction(&txn, None, None)
+                {
+                    Ok(response) => response.into_inner(),
+                    Err(WaitForTransactionError::TransactionHashMismatchError(_))
+                    | Err(WaitForTransactionError::TransactionExpired) => continue 'faucet,
```

Comment:
So  Faucet returns a list of txns because the way it works is 1) issue txn for creating account 2) issue txn to fund said account. And both of them need to complete in order to have a functioning account with some balance. There shouldn't ever be an instance where we accidently mint too much because if we actually have the mint txn successfully complete this function will return.

Reply:
got it that makes sense. in which case, this is probably fine